### PR TITLE
handling error when member left guild

### PR DESCRIPTION
**Pull request by ankitsmt211**
During auto-archive routine for help threads, bot sends an embed educating OP why the thread was archived, it also pings OP in the process. This whole archiving routine depends on whether member was retrieved, thus it fails consecutively until discords built in archive feature kicks in.

image

### CODE REVIEW

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:135]: if something breaks, should we archive anyway?

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:151]: just say cannot mention user

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:135]: yes ig

**Review by Zabuzard — APPROVED** [review]

**Review by Zabuzard — APPROVED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:160]: do an `onErrorMap` so we'll always only do 1 queue

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:160]: not sure if i understand wym, could you show an example? Thanks

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:160]:
```java
        .flatMap(...)
        .onErrorMap(any -> {
            log...
            return threadChannel.archive();
        })
        .queue();
```
Something like  this

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:160]: (or onErrorFlatMap, if that results in another rest action)

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:161]: too ugly?

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:147]: lol whole gang is here

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:162]: better to  make put the code here instead of making  consumer

**Review by Taz03 — CHANGES_REQUESTED** [review]

**Review by Zabuzard — CHANGES_REQUESTED** [review]
that duplication here isnt good:
dup

also, ur technically also attempting to archive when archiving itself failed, which isnt good.
the "try this, but always archive" shouldnt span across archiving itself.

instead, try to write it like:
```java
...
.flatMap(member -> ...)
.onError(handleFailure)
.flatMap(any -> threadChannel.getManager().setArchived(true))
.queue();
```

**Comment by ankitsmt211** [comment]
@Zabuzard im blank here, it requires me to return a `RestAction<Message>`, i don't think i can do that with `OnError....` since we already handled with or without mention in `mapToResult`, there's no point in sending a new message. The only reason i was think to handle archiving as backup was if something breaks and `AutoArchiver` retries again and again so archive anyway. On a second note, do we even need it? `Threads` cant be deleted, only other variable was member leaving. Now if somehow it breaks then during second try, should be archived anyway. Thoughts?

image

**Comment by Zabuzard** [comment]
might misremember the name. onFailure(...) perhaps.
personally, i dont mind u removing the error handling on that part. i dont think it's necessary

**Comment by Zabuzard** [comment]
just checked. was looking for the equivalent to my onSuccess(...) method. seems it doesnt exist.
so it would either be:

.onErrorMap(any -> {
  do the log thing;
  return any;
}

or via mapToResults

**Comment by ankitsmt211** [comment]
seems much better, at least chain is pretty readable now.

**Review by Zabuzard — DISMISSED** [review]

**Comment by ankitsmt211** [comment]
Don't merge yet.

**Comment by ankitsmt211** [comment]
@Taz03 It's good to go when you can.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:166]:
```suggestion
                if (sentEmbed.getFailure() instanceof ErrorResponseException) {
                    // log
                }

                return threadChannel.getManager().setArchived(true);
```

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:166]: what's the issue with earlier implementation? It tells a beautiful story, if embed is a success archive the thread otherwise handle failure..which logs and then regardless archives it. It looks like you're trying to squeeze in the failure part at the cost of readability.

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:166]: if its failure log, and archive anyways. doesn't need to be more complex than that

Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:166]: and it's doing that exactly, what im trynna say is you're squeezing in the failure part. Im trying to clean the chain cleaner, like i said it's already doing too much.

Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:166]: The functional implementation is local to method, so you dont have to jump around. It has a name as well, which abstracts away the details when reading pipeline. So something bad happened, oki it's calling a function `handleFailure` which consumes a `throwable` so maybe ill look at later but ik it handles failures. 

I don't wanna do the if check inline because, it just adds more brackets then you're keeping track of what's what inside chain. Ill do the changes but i still feel the same way.

**Review by Taz03 — DISMISSED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (surajkumar) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:149]: Do we want to only do this on `ErrorResponseException`? What about any other type of error, they should be logged too.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (surajkumar) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:144]: As discussed on the server already, we should add more context to this. I would consider this a `warning` but if this happens all the time, can we change this bit of code so it's clear what's going on e.g. adding a check so see if the user is in server, if they are, error/warn log this message. Otherwise, info log that they have left.

Comment (surajkumar) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:144]: Also, constants should be uppercase. As should the `Logger`, so make it `LOGGER` while you're here please. It's not ugly to make it LOGGER.

**Review by surajkumar — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:149]: I suppose we can log it, goal was to handle errors inside JDA anything over network or external is logged by JDA no(correct me if i'm wrong)?

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:144]: since we already handled this with `sendEmbedWithoutMention`, it just had to be there could be in debug or info as long as it's taken care of.

Comment (surajkumar) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:144]: The problem is that you're needing to explain this. At the minimum add documentation to anybody else without context will know what's going on. But I'd much rather prefer the code to be rewritten in a way that can express it. You can choose the solution here either way.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (surajkumar) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:149]: Maybe but if we're being given a `getFailure()` we should use it.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java]: I have made some changes in pipeline to display intent more clearly, lmk what you think.

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (surajkumar) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:139]: Nice good readabilty change

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java]: this is valid but everywhere else in the code, it's `logger`, so for consistency, keep it `logger`. And if you want to  change it,  change it everywhere else as well in a separate pr

Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java]: i was wondering why i used `logger` twice in both PRs, prolly followed from rest of code. Can we start from here? It prolly should be done anyway.

Comment (surajkumar) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java]: If it's lowercase logger everywhere then yeah it should be changed. Surprised the linter doesn't flag it. But changing it everywhere is out of the scope of this PR. 

We have 2 solutions:

1. Create a PR to address this 
2. Update whenever we touch a file

I'd go for the second option, if we see such cases, then they can be fixed when next touched. 

We can raise a request for somebody to do it in a big PR though for those who want to contribute but don't understand enough. 

In any case, it's good how it is now in this PR with the uppcase.

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java]: > In any case, it's good how it is now in this PR with the uppcase.

its not, no one has agreed to the change + it'll only make the code more inconsistent

Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java]: changed back to `logger`, new PR can be raised for refactoring later.

**Review by Taz03 — APPROVED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:155]: well, in what case can embed sending fail? now that u handle user not fetchable separately. Ig we can safely remove this part

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java
Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:155]: it's supposed to be catch it all, just to be at least aware if something fails on sending message that can be handled in future. Are you sure?

Comment (Taz03) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:155]: well, we never check if a message sending failed so... remove it

Comment (ankitsmt211) [reply application/src/main/java/org/togetherjava/tjbot/features/help/HelpThreadAutoArchiver.java:155]: fair

**Review by Taz03 — APPROVED** [review]

**Review by marko-radosavljevic — APPROVED** [review]
Thanks. :heart:

**Pull request merged**
